File: LeetCode_Algorithm/589/589.py
from typing import List
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = []

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        # O(n) 56ms 89%
        rootnode = Node(1)
        a_node = Node(2); b_node = Node(3); c_node = Node(4)
        rootnode.children.append(a_node);
        rootnode.children.append(b_node);
        rootnode.children.append(c_node);
        return self.solve(rootnode, [])

    def solve(self, root, ans):
        if not root:
            return ans
        ans.append(root.val)
        for child in root.children:
            self.solve(child, ans)
        return ans
        
        

        
        

            
            

print(Solution().preorder([1,None,3,2,4,None,5,6]))

File: LeetCode_Algorithm/589/589.js
function Node(val, children) {
   this.val = val;
   this.children = []
};

function solve(root, ans) {
  if (!root) {
    return ans
  }
  ans.push(root.val)
  for (let child of root.children) {
    solve(child, ans)
  }
  return ans
};

var preorder = function() {
  // O(n) 91ms 82%
  rootNode = new Node(1)
  aNode = new Node(2); bNode = new Node(3); cNode = new Node(4);
  rootNode.children.push(aNode);
  rootNode.children.push(bNode);
  rootNode.children.push(cNode);

  return solve(rootNode, [])
};

console.log(preorder())